services/src/main/java/io/scalecube/services/RemoteServiceInstance.java

```diff
@@ -0,0 +1,163 @@
+package io.scalecube.services;
+
+import io.scalecube.cluster.ICluster;
+import io.scalecube.transport.Address;
+import io.scalecube.transport.Message;
+
+import com.google.common.base.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+public class RemoteServiceInstance implements ServiceInstance {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RemoteServiceInstance.class);
+
+  private final ICluster cluster;
+  private final Address address;
+  private final String memberId;
+  private final String serviceName;
+
+  /**
+   * Remote service instance constructor to initiate instance.
+   * 
+   * @param cluster to be used for instance context.
+   * @param serviceReference service reference of this instance.
+   */
+  public RemoteServiceInstance(ICluster cluster, ServiceReference serviceReference) {
+    this.serviceName = serviceReference.serviceName();
+    this.cluster = cluster;
+    this.address = serviceReference.address();
+    this.memberId = serviceReference.memberId();
+  }
+
+  @Override
+  public String serviceName() {
+    return serviceName;
+  }
+
+  @Override
+  public Object invoke(Message request, ServiceDefinition definition) throws Exception {
+    Preconditions.checkArgument(definition != null, "Service definition can't be null");
+
+    // Try to call via messaging
+    // Request message
+    if (definition.returnType().equals(CompletableFuture.class)) {
+      if (definition.parametrizedType().equals(Message.class)) {
+        return futureInvokeMessage(request);
+      } else {
+        return futureInvokeGeneric(request);
+      }
+    } else {
+      CompletableFuture<Object> future = futureInvokeGeneric(request);
+      Object result = future.get();
```

Comment:
I think this call is dangerous, we should not ever wait forever. Maybe even we should not allow to do sync call, or rethink implementation of it. For example encapsulate method call in function and delay call to this function until we receive result from remote node @antonkharenko, @ronenhamias WDYT?